subsys/bluetooth/mesh/scene_srv.c

```diff
@@ -0,0 +1,716 @@
+}
+
+static void handle_store_unack(struct bt_mesh_model *mod,
+			       struct bt_mesh_msg_ctx *ctx,
+			       struct net_buf_simple *buf)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+
+	(void)scene_store(srv, net_buf_simple_pull_le16(buf));
+}
+
+static void handle_delete(struct bt_mesh_model *mod,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+	enum bt_mesh_scene_status status;
+	uint16_t *scene;
+
+	scene = scene_find(srv, net_buf_simple_pull_le16(buf));
+	if (!scene) {
+		status = BT_MESH_SCENE_NOT_FOUND;
+	} else {
+		scene_delete(srv, scene);
+		status = BT_MESH_SCENE_SUCCESS;
+	}
+
+	scene_register_status_send(srv, ctx, status);
+}
+
+static void handle_delete_unack(struct bt_mesh_model *mod,
+				struct bt_mesh_msg_ctx *ctx,
+				struct net_buf_simple *buf)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+	uint16_t *scene;
+
+	scene = scene_find(srv, net_buf_simple_pull_le16(buf));
+	if (scene) {
+		scene_delete(srv, scene);
+	}
+}
+
+const struct bt_mesh_model_op _bt_mesh_scene_setup_srv_op[] = {
+	{
+		BT_MESH_SCENE_OP_STORE,
+		BT_MESH_SCENE_MSG_LEN_STORE,
+		handle_store,
+	},
+	{
+		BT_MESH_SCENE_OP_STORE_UNACK,
+		BT_MESH_SCENE_MSG_LEN_STORE,
+		handle_store_unack,
+	},
+	{
+		BT_MESH_SCENE_OP_DELETE,
+		BT_MESH_SCENE_MSG_LEN_DELETE,
+		handle_delete,
+	},
+	{
+		BT_MESH_SCENE_OP_DELETE_UNACK,
+		BT_MESH_SCENE_MSG_LEN_DELETE,
+		handle_delete_unack,
+	},
+	BT_MESH_MODEL_OP_END,
+};
+
+static int scene_srv_pub_update(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+
+	scene_status_encode(srv, &srv->pub_msg, BT_MESH_SCENE_SUCCESS);
+	return 0;
+}
+
+static int scene_srv_init(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+
+	if (mod->id == BT_MESH_MODEL_ID_SCENE_SETUP_SRV) {
+		srv->setup_mod = mod;
+		return 0;
+	}
+
+	srv->mod = mod;
+	net_buf_simple_init_with_data(&srv->pub_msg, srv->buf,
+				      sizeof(srv->buf));
+	srv->pub.msg = &srv->pub_msg;
+	srv->pub.update = scene_srv_pub_update;
+	return 0;
+}
+
+static int scene_srv_set(struct bt_mesh_model *mod, const char *path,
+			 size_t len_rd, settings_read_cb read_cb, void *cb_arg)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+	uint8_t buf[SCENE_PAGE_SIZE];
+	uint16_t scene;
+	ssize_t size;
+	uint8_t page;
+	bool vnd;
+
+	BT_DBG("path: %s", log_strdup(path));
+
+	if (!strcmp(path, "s")) {
+		(void)read_cb(cb_arg, &srv->curr, sizeof(srv->curr));
+		BT_DBG("Current scene 0x%x", srv->curr);
+		return 0;
+	}
+
+	scene = strtol(path, NULL, 16);
+	if (!scene) {
+		BT_ERR("Unknown data %s", log_strdup(path));
+		return 0;
+	}
+
+	settings_name_next(path, &path);
+	if (!path) {
+		return 0;
+	}
+
+	vnd = path[0] == 'v';
+	page = strtol(&path[1], NULL, 16);
+	if (vnd) {
+		srv->vndpages = MAX(page, srv->vndpages);
+	} else {
+		srv->sigpages = MAX(page, srv->sigpages);
+	}
+
+	if (!bt_mesh_is_provisioned()) {
```

Comment:
Is this possible? Aren't we suppose to have _clean_ device if it is not provisioned or am I missing something?